atd-etl/data_model/utils/process_pdfs.py

```diff
@@ -0,0 +1,174 @@
+
+    if s3_upload:
+        s3_object_key_pdf = get_cr3_object_key(filename, "pdfs")
+        logger.debug(f"Uploading CR3 pdf to {s3_object_key_pdf}")
+        upload_file_to_s3(pdf_path, s3_object_key_pdf)
+
+        s3_object_key_diagram = get_cr3_object_key(diagram_filename, "crash_diagrams")
+        logger.debug(f"Uploading crash diagram to {s3_object_key_diagram}")
+        upload_file_to_s3(diagram_full_path, s3_object_key_diagram)
+
+        logger.debug(f"Updating crash CR3 metadata")
+        res = make_hasura_request(
+            query=UPDATE_CRASH_CR3_FIELDS,
+            variables={
+                "crash_id": crash_id,
+                "data": {
+                    "cr3_processed_at": datetime.now(timezone.utc).isoformat(),
+                    "cr3_stored_fl": True,
+                },
+            },
+        )
+        # check to make sure that we actually updated a crash record
+        affected_rows = res["update_crashes_cris"]["affected_rows"]
+        if not affected_rows:
+            raise ValueError(
+                f"Crash ID: {crash_id} - CR3 PDF has no matching crash record in the DB. This should never happen."
+            )
+
+
+def process_pdfs(extract_dir, s3_upload, max_workers):
+    """Main loop for extract crash diagrams from  CR3 PDFs
+
+    Args:
+        extract_dir (str): the local path to the current extract
+        s3_upload (bool): if the diagram and PDF should be uploaded to the S3 bucket
+        max_workers (int): the maximum number of workers to assign to the
+            multiprocressing pool
+    """
+    overall_start_tme = time.time()
+    # make the crash_diagram extract directory
+    Path(os.path.join(extract_dir, "crash_diagrams")).mkdir(parents=True, exist_ok=True)
+
+    pdfs = [
+        filename
+        for filename in os.listdir(os.path.join(extract_dir, "crashReports"))
+        if filename.endswith(".pdf")
+    ]
+    pdf_count = len(pdfs)
+    logger.info(f"Found {pdf_count} PDFs to process")
+
+    futures = []
+    with ProcessPoolExecutor(max_workers=max_workers) as executor:
+        for index, filename in enumerate(pdfs):
+            future = executor.submit(
+                process_pdf, extract_dir, filename, s3_upload, index
+            )
+            futures.append(future)
+    # inspect results and collect all errors—this must be done after all pdfs have
+    # processed because future.result() is blocking
+    errors = []
+    for index, future in enumerate(futures):
+        try:
+            future.result()
+        except Exception as e:
+            # grab the filename from the pdf list
+            filename = pdfs[index]
+            errors.append([filename, e])
+
+    if errors:
+        logger.error(
+            f"Encountered {len(errors)} error(s) processing PDFs. Logging up to 10 errors and raising the first..."
+        )
+        for filename, e in errors[0:10]:
+            logger.info(f"Error processing {filename}: {e}")
+        raise errors[0][1]
```

Comment:
because of concurrency, we will not know if the PDF processing is bombing until _all_ PDFs have been processed.

this isn't a major concern for processing a normal-sized extract of ~200 crashes, but really not ideal if we were processing thousands of PDFs. like would take us a few hours to find out we had a misconfigured env var.

i didn't want to over-engineer this off the bat, but one improvement we could make would be to process the PDFs in chunks of 100, inspecting each chunk for errors before running another batch.

Reply:
👍 will open a follow-up issue